### PR TITLE
[Feature] Callbacks, focus management, copy-paste action

### PR DESCRIPTION
**Pull request by jorGonzalez291292**
## Useful links
- [Contributing](https://github.com/mercadolibre/fury_andesui-android/blob/develop/CONTRIBUTING.md) has more information and tips for a great pull request. This will give you a nice introduction about how to contribute with Andes.
- [Wiki](https://github.com/mercadolibre/fury_andesui-android/wiki) has all the info you need to start with Andes. Even I heard there's a video there...

## Motivation
Create a new component that allows us to write a code.

## Description
Callbacks are added to listen to the changes in the textfield code, handling of focus when you write and delete. Copy and paste action.
Try to answer those questions:
- What is it about?
- How does it work?

## Affected component
AndesTextfield is affected.

## Frontify component link
The UX team behind Andes keeps all the Andes Components inside the Frontify site. Each component has its own section. Let us know the section of the component you want to modify.

## Screenshots

[Api19](https://i.imgur.com/W0dcgXc.mp4)
[Api29](https://i.imgur.com/krZVYTK.mp4)

## Checks
Are you sure you followed all those steps? In such case, let's say with me "I solemnly declare that ...":
- [ ] I have coded an example inside the Demo App,
- [ ] I have added proper tests,
- [ ] I have modified the [Changelog](https://github.com/mercadolibre/fury_andesui-android/blob/develop/CHANGELOG.md) file,
- [ ] I have updated GitHub wiki,
- [ ] I have the explicit approval of one or more members of the UX Team,
- [ ] I have updated the version in gradle.properties file.

## Change type
This is easy, let us know what this code is about:
- [ ] This code adds a new component
- [ ] This code includes improvements to an existent component
- [ ] This code improves or modifies ONLY the Demo App.

## Check common errors
Whether you are an Android or iOS developer and if you're still there then we'll give you a present:
https://proandroiddev.com/how-to-maximize-androids-ui-reusability-5-common-mistakes-cb2571216a9f

### CODE REVIEW

**Review by jperezgonzal — APPROVED** [review]

File: components/src/main/java/com/mercadolibre/android/andesui/textfield/AndesEditText.kt [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+ */
+internal class AndesEditText : AppCompatEditText {
+
+    private var contextMenuListener: OnTextContextMenuItemListener = object : OnTextContextMenuItemListener {}
+
+    constructor(context: Context) : super(context)
+    constructor(context: Context, attrs: AttributeSet?) : super(context, attrs)
+    constructor(context: Context, attrs: AttributeSet?, defStyleAttr: Int) : super(context, attrs, defStyleAttr)
+
+    fun setOnTextContextMenuItemListener(contextMenuListener: OnTextContextMenuItemListener) {
```
Comment (haurbano) [inline components/src/main/java/com/mercadolibre/android/andesui/textfield/AndesEditText.kt:19]: **NO** es necesario cambiarlo pero personalmente me gusta la practica de tener una lista de listeners y no un solo, asi varios pueden escuchar este callback si estan interesados. Igual para la proxima si te gusta la idea, por ahora esta perfecto! 😄

File: components/src/main/java/com/mercadolibre/android/andesui/textfield/AndesTextfieldCode.kt [hunk 2]
```diff
@@ -292,11 +430,19 @@ class AndesTextfieldCode : ConstraintLayout {
     }
 
     private fun getBoxAt(index: Int): AndesTextfield? {
-        return textfieldBoxCodeContainer.getChildAt(index) as AndesTextfield?
+        return textfieldBoxCodeContainer.getChildAt(index) as? AndesTextfield?
```
Comment (haurbano) [inline components/src/main/java/com/mercadolibre/android/andesui/textfield/AndesTextfieldCode.kt]: Qué pasaria si no es un AndesTextField? Podriamos lanzar un error en caso de que no lo sea? Asi alertamos al dev que debe ser un AndesTextField. Por ejemplo:
`throw CustomException("Shoudl be a AndesTextField")`

File: demoapp/src/main/java/com/mercadolibre/android/andesui/demoapp/feature/TextfieldShowcaseActivity.kt [hunk 3]
```diff
@@ -284,13 +285,35 @@ class TextfieldShowcaseActivity : AppCompatActivity() {
             stateAdapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item)
             styleSpinner.adapter = styleAdapter
 
+            textfieldCode.setOnTextChangeListener(object : AndesTextfieldCode.OnTextChangeListener {
+                override fun onChange(text: String) {
+                    Log.i("JORGE", "TEXT CHANGE: $text")
```
Comment (haurbano) [inline demoapp/src/main/java/com/mercadolibre/android/andesui/demoapp/feature/TextfieldShowcaseActivity.kt]: Se nos fue el log? O cual es la razon del tag "JORGE"? Pregunto no mas 😄

File: components/src/main/java/com/mercadolibre/android/andesui/textfield/AndesTextfieldCode.kt (hunk 2 above)
Comment (jorGonzalez291292) [reply components/src/main/java/com/mercadolibre/android/andesui/textfield/AndesTextfieldCode.kt]: Realmente no debería pasar ya que internamente estoy agregando Views de tipo `AndesTextfield`, y el integrador no tiene control de esto. La verdad el check `as?` podría estar de más.

File: demoapp/src/main/java/com/mercadolibre/android/andesui/demoapp/feature/TextfieldShowcaseActivity.kt (hunk 3 above)
Comment (jorGonzalez291292) [reply demoapp/src/main/java/com/mercadolibre/android/andesui/demoapp/feature/TextfieldShowcaseActivity.kt]: jajaja Lo cambio a "ANDES".

File: components/src/main/java/com/mercadolibre/android/andesui/textfield/AndesEditText.kt (hunk 1 above)
Comment (jorGonzalez291292) [reply components/src/main/java/com/mercadolibre/android/andesui/textfield/AndesEditText.kt:19]: Dale, gracias por el consejo!

File: components/src/main/java/com/mercadolibre/android/andesui/textfield/AndesTextfieldCode.kt (hunk 2 above)
Comment (haurbano) [reply components/src/main/java/com/mercadolibre/android/andesui/textfield/AndesTextfieldCode.kt]: Bueno en ese caso como tu consideres, si no es necesario el `as?` entonces dejemolo como `as`. Asi la funcion puede devolver un `AndesTextField` en lugar de un `AndesTextField?`

**Review by haurbano — APPROVED** [review]

**Pull request merged**